sdk/attestation/azure-security-attestation/src/attestation_client.cpp

```diff
@@ -0,0 +1,158 @@
+      options.InittimeData,
+      options.RuntimeData,
+      options.DraftPolicyForAttestation,
+      options.Nonce};
+
+  std::string serializedRequest(AttestSgxEnclaveRequestSerializer::Serialize(attestRequest));
+
+  auto encodedVector = std::vector<uint8_t>(serializedRequest.begin(), serializedRequest.end());
+  Azure::Core::IO::MemoryBodyStream stream(encodedVector);
+  auto request = AttestationCommonRequest::CreateRequest(
+      m_endpoint, m_apiVersion, HttpMethod::Post, {"attest/SgxEnclave"}, &stream);
+
+  // Send the request to the service.
+  auto response = AttestationCommonRequest::SendRequest(*m_pipeline, request, context);
+
+  // Deserialize the Service response token and return the JSON web token returned by the service.
+  std::string responseToken = AttestationServiceTokenResponseSerializer::Deserialize(response);
+
+  // Parse the JWT returned by the attestation service.
+  auto token
+      = AttestationTokenInternal<AttestationResult, AttestationResultSerializer>(responseToken);
+
+  // Validate the token returned by the service. Use the cached attestation signers in the
+  // validation.
+  CacheAttestationSigners(context);
+  token.ValidateToken(
+      options.TokenValidationOptions.HasValue() ? options.TokenValidationOptions.Value()
+                                                : this->m_tokenValidationOptions,
+      m_attestationSigners);
+
+  // And return the attestation result to the caller.
+  auto returnedToken = AttestationToken<AttestationResult>(token);
+  return Response<AttestationToken<AttestationResult>>(returnedToken, std::move(response));
+}
+
+Azure::Response<AttestationToken<AttestationResult>> AttestationClient::AttestOpenEnclave(
+    std::vector<uint8_t> const& openEnclaveReport,
+    AttestOptions options,
+    Azure::Core::Context const& context) const
+{
+  AttestOpenEnclaveRequest attestRequest{
+      openEnclaveReport,
+      options.InittimeData,
+      options.RuntimeData,
+      options.DraftPolicyForAttestation,
+      options.Nonce};
+  std::string serializedRequest(AttestOpenEnclaveRequestSerializer::Serialize(attestRequest));
+
+  auto encodedVector = std::vector<uint8_t>(serializedRequest.begin(), serializedRequest.end());
+  Azure::Core::IO::MemoryBodyStream stream(encodedVector);
+  auto request = AttestationCommonRequest::CreateRequest(
+      m_endpoint, m_apiVersion, HttpMethod::Post, {"attest/OpenEnclave"}, &stream);
+
+  auto response = AttestationCommonRequest::SendRequest(*m_pipeline, request, context);
+  std::string responseToken = AttestationServiceTokenResponseSerializer::Deserialize(response);
+  auto token
+      = AttestationTokenInternal<AttestationResult, AttestationResultSerializer>(responseToken);
+  return Response<AttestationToken<AttestationResult>>(token, std::move(response));
+}
+
+void AttestationClient::CacheAttestationSigners(Azure::Core::Context const& context) const
+{
+  std::unique_lock<std::shared_timed_mutex> stateLock(m_sharedStateLock);
```

Comment:
I'm confused. This impl *is* in a cpp file. Where should I be moving it to?

Reply:
my bad.

Reply:
However, you can create the mutex within this cpp file ( use unamed namespace ) and then remove the mutex field from the the class in the hpp

Reply:
How do I scope that unnamed namespace mutex to the specific instance of the attestation client? Otherwise we serialize all clients on the mutex, which can lead to boxcar issues (maybe not a huge deal in practice, but I wouldn't want to use that as an examplar of how to design multi-threaded access).

Reply:
The `mutable` storage class indicates that a member may be changed when called from a const function. And `CacheAttestationSigners` is marked as const because it's called from `const` functions.

In general, I strongly prefer that locks be scoped as tightly as possible to their use case. Is there a reason you want this lock to be global? Is it solely to avoid having the `#include <mutex>`? If so, what are the rules w.r.t. standard headers? Which ones should be avoided? Encouraged?